### PR TITLE
Partially revert #32137

### PR DESCRIPTION
**Pull request by haampie**
PR #32137 was based on the wrong assumption that direct fetches from
mirrors are necessarily slow.

The solution (fetch the index to locally see if something is in it) is
not scalable.

Instead, there's plenty of space to improve the current s3:// fetch
logic: it shouldn't create a client every time, the number of 404s can
be halved through #34325, etc.

Also, there's a work around for some of the slowness, since aws
credentials can be stored in ~/.aws/credentials or env variables,
which avoid online authentication with amazon (which presumably
generates some temporary access token).

### CODE REVIEW

**Comment by blue42u** [comment]
Spinning out a few notes from [our previous discussion](https://github.com/spack/spack/pull/32137#issuecomment-1336810056):

Direct fetches are ridiculously expensive from an S3 URL, they take about 15 seconds per mirror per package in the spec:
```console
$ spack -dd install --cache-only zlib
...
==> [2022-12-05-08:50:37.938981, 375704] WRITE LOCK: .../cache/indices/.contents.json.lock[0:0] [Acquired at 08:50:37.938969] 
==> [2022-12-05-08:50:37.939310, 375704] WRITE LOCK: .../cache/indices/.contents.json.lock[0:0] [Released at 08:50:37.939295] 
==> [2022-12-05-08:50:53.016823, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on s3://spack-binaries/does/not/exist/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json.sig
  Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-12-05-08:50:53.016893, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on s3://spack-binaries/does/not/exist/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json
  Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-12-05-08:50:53.016917, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on s3://spack-binaries/does/not/exist/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.yaml
  Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-12-05-08:50:53.025522, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on file:///home/deepthought/.spack-manual/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json.sig
  Download failed: <urlopen error [Errno 2] No such file or directory: '/home/deepthought/.spack-manual/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json.sig'>
==> [2022-12-05-08:50:53.025571, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on file:///home/deepthought/.spack-manual/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json
  Download failed: <urlopen error [Errno 2] No such file or directory: '/home/deepthought/.spack-manual/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json'>
==> [2022-12-05-08:50:53.025593, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on file:///home/deepthought/.spack-manual/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.yaml
  Download failed: <urlopen error [Errno 2] No such file or directory: '/home/deepthought/.spack-manual/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.yaml'>
==> [2022-12-05-08:50:54.469689, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on https://mirror.spack.io/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json.sig
  Download failed: HTTP Error 404: Not Found
==> [2022-12-05-08:50:54.469763, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on https://mirror.spack.io/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json
  Download failed: HTTP Error 404: Not Found
==> [2022-12-05-08:50:54.469784, 375704] Did not find linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.json on https://mirror.spack.io/build_cache/linux-debiann-zen2-gcc-12.2.0-zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt.spec.yaml
  Download failed: HTTP Error 404: Not Found
==> [2022-12-05-08:50:54.473964, 375704] Error: No binary for zlib-1.2.13-ukupt3qvrkh2ic4ve4c4jmkqaumcowqt found when cache-only specified
```
So for a Spec of 50 packages, every mirror that can't contribute generates 12.5 minutes of waste time.

Using global AWS credentials is a security problem and should never be done, for the reasons I mention in bullet 3 of [this discussion](https://github.com/spack/spack/discussions/32958). Note also that (in my observations) using AWS credentials only decreases 15 seconds to about 3 (for 50 packages, 2.5 min per mirror). It does **not** solve the fundamental problem.

However, the solution implemented in #32137 is not scalable in `spack ci` build jobs because of other problems, two of which are algorithmic in nature:
 1. The code in [`BinaryCacheIndex._fetch_and_cache_index`](https://github.com/spack/spack/blob/76417d6ac6679ac65cfe619f0f5ec0b3175647d2/lib/spack/spack/binary_distribution.py#L424) is improperly parallelized, multiple parallel Spacks will (attempt to) fetch the huge `index.json` instead of just one, and
 2. Each build job must "identify" which mirror to fetch each package from which requires `O(# packages * # mirrors)` time/requests per job (without a shared index cache), instead of deterministically planning in `spack ci generate` which would only require `O(# packages)` time/requests in each build job.

The root issues will take more time and effort to repair, this much smaller reversion fixes the slowdowns in Spack CI caused by re-fetching the 172MB `index.json` many times.

**Comment by haampie** [comment]
You don't have to expose a token with write access; as I said it's a workaround, necessarily until the client object thing is instantiated once and only once. Cherry-pick #34325 for 50% fewer fewer GET requests on 404. Exchange inner and outer loop in the binary cache fetch logic to save another 66% (outer loop: .sig, .json. yaml; inner loop: mirrors) on success.

**Comment by blue42u** [comment]
> Exchange inner and outer loop in the binary cache fetch logic to save another 66% (outer loop: .sig, .json. yaml; inner loop: mirrors).

This will increase the fetch logic by `O(# mirrors)` for packages that can only be fetched from an unsigned buildcache (which is all of them in our CI because the Spack public buildcache doesn't currently support many of the OSs we use).

**Comment by haampie** [comment]
Then use https:// for fetch instead of s3:// until things are fixed :man_shrugging:

**Comment by blue42u** [comment]
We can't use `https://` for fetch because our bits are in a private S3 bucket, we need S3 authentication to access it. :facepalm:

**Comment by haampie** [comment]
Using #34325 + #34324 + a small diff to not create a client on every request, I'm getting:

```
$ spack mirror list
wut             s3://spack-binaries (fetch)
wut             s3://spack-binaries (push)
spack-public    https://mirror.spack.io
```

```python
In [1]: import spack.binary_distribution

In [2]: s = spack.spec.Spec("zlib").concretized()

In [3]: %timeit spack.binary_distribution.try_direct_fetch(s)
739 ms ± 6.29 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

for a miss, which is good enough.

Compared to develop:

```python
In [1]: import spack.binary_distribution

In [2]: s = spack.spec.Spec("zlib").concretized()

In [3]: %timeit spack.binary_distribution.try_direct_fetch(s)
15.7 s ± 99.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
---
> Using global AWS credentials is a security problem and should never be done

> we need S3 authentication to access it

:man_shrugging:

**Comment by blue42u** [comment]
Fine, valid point. Once the other two PRs land I can adjust our CI accordingly.

I'm still going to make the algorithmic adjustments mentioned previously, once I find a bit of time.

**Review by blue42u — APPROVED** [review]
@tldahlgren I'm not a Spack dev, so... I guess I'll just :+1: this as the previous author for the removed code?

The improvements I want to make will take more time, and this PR un-breaks an issue that occurs in Spack CI today, so I'm good with merging this and cleaning up in a later PR.

**Comment by tldahlgren** [comment]
> @tldahlgren I'm not a Spack dev, so... I guess I'll just 👍 this as the previous author for the removed code?
> 
> The improvements I want to make will take more time, and this PR un-breaks an issue that occurs in Spack CI today, so I'm good with merging this and cleaning up in a later PR.

Perhaps I should start adding a comment to PRs when I "assign" it to someone who has commented.  Doing so is an informal way of flagging who -- besides spackbot -- has commented on a PR so it shows up on the PR list page.

**Comment by blue42u** [comment]
@tldahlgren Ah, got it. In our repository those "assigned" to a PR are roughly those responsible for testing/merging it (sometimes, we're a small team and a bit loose with it). So I thought I was being requested to approve/merge it or something. Sorry for the confusion!

File: lib/spack/spack/binary_distribution.py
Comment (scottwittenburg) [inline lib/spack/spack/binary_distribution.py:270]: Isn't this already part of #34359?

Comment (haampie) [reply lib/spack/spack/binary_distribution.py:270]: Yes. Will close this now

**Comment by haampie** [comment]
Closed in favor of #34359

**Pull request closed without merging**
